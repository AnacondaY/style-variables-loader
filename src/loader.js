import { getOptions } from 'loader-utils';
import glob from 'glob';
import convert from './convertor';

const defaultOptions = {
    style: 'scss',
    files: [],
    variables: {},
    hot: false,
    cwd: process.cwd()
}

export default function(source){
    this.cacheable();
    let options = getOptions(this);
    options = {
        ...defaultOptions,
        ...options
    }
    let { style, files, cwd, variables, hot } = options;
    if(typeof files === 'string'){
        files = glob.sync(files, {
            cwd,
            absolute: true
        });
    }
    const vars = files.reduce((vars, file) => {
        const data = require(file);
        return Object.assign(vars, data);
    }, {});
    if(hot){
        files.forEach(file => {
            this.addDependency(file);
        });
    }
    const result = Object.assign(vars, variables);
    return `${convert(result, style)}\n${source}`;
}