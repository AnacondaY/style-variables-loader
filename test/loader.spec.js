import path from 'path';
import loader from '../src/loader';

const context = {
    cacheable: jest.fn(),
    addDependency: jest.fn(),
};
const mockDir = path.resolve(__dirname, 'mocks');
let mockOptions;

jest.mock('loader-utils', () => ({
    getOptions: () => mockOptions,
}));

describe('=== loader test ===', () => {
    let _loader;

    beforeAll(() => {
        _loader = (options) => {
            mockOptions = options;
            return loader.call(context, 'some style content');
        };
    });

    describe('inline variables', () => {
        let variables;

        beforeEach(() => {
            variables = {
                var1: 'red',
                var2: 'navy',
            };
        });

        test('scss', () => {
            const result = _loader({ variables, style: 'scss' });
            expect(result).toMatchSnapshot();
        });

        test('sass', () => {
            const result = _loader({ variables, style: 'sass' });
            expect(result).toMatchSnapshot();
        });

        test('less', () => {
            const result = _loader({ variables, style: 'less' });
            expect(result).toMatchSnapshot();
        });

        test('stylus', () => {
            const result = _loader({ variables, style: 'stylus' });
            expect(result).toMatchSnapshot();
        });
    });

    describe('input files', () => {
        // test('from js', () => {
        //     const result = _loader({
        //         files: [
        //             path.resolve(mockDir, 'fromJs.js'),
        //         ]
        //     });

        //     expect(result).toMatchSnapshot();
        // });

        test('from json', () => {
            const result = _loader({
                files: [
                    path.resolve(mockDir, 'fromJSON.json'),
                ] 
            });

            expect(result).toMatchSnapshot();
        });

        // test('both types', () => {
        //     const result = _loader({ files: './test/mocks/*.*' });
        //     expect(result).toMatchSnapshot();
        // });
    });

    // describe('hot load', () => {
    //     const files = [
    //         path.resolve(mockDir, 'fromJs.js'),
    //     ];

    //     beforeAll(() => {
    //         _loader({ files, hot: true });
    //     });

    //     test('run by files change', () => {
    //         expect(context.addDependency).toHaveBeenCalledTimes(files.length);
    //         files.forEach((file) => {
    //             expect(context.addDependency).toBeCalledWith(file);
    //         });
    //     });
    // });
});
