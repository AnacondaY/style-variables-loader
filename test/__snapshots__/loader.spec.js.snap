// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`=== loader test === inline variables less 1`] = `
"@var1:red;
@var2:navy;
some style content"
`;

exports[`=== loader test === inline variables sass 1`] = `
"$var1:red
$var2:navy
some style content"
`;

exports[`=== loader test === inline variables scss 1`] = `
"$var1:red;
$var2:navy;
some style content"
`;

exports[`=== loader test === inline variables stylus 1`] = `
"var1=red
var2=navy
some style content"
`;

exports[`=== loader test === input files both types 1`] = `
"$var1:red;
$var2:navy;
some style content"
`;

exports[`=== loader test === input files from js 1`] = `
"$var1:red;
$var2:nav;
some style content"
`;

exports[`=== loader test === input files from json 1`] = `
"$var1:red;
$var2:navy;
some style content"
`;
